File: projects/vpn-software/src/packet.rs
//! Packet parsing and construction for VPN protocol

use std::net::{Ipv4Addr, Ipv6Addr};

use crate::error::{Result, VpnError};

/// IP protocol numbers
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
#[repr(u8)]
pub enum IpProtocol {
    ICMP = 1,
    TCP = 6,
    UDP = 17,
    ICMPv6 = 58,
    Other(u8),
}

impl From<u8> for IpProtocol {
    fn from(value: u8) -> Self {
        match value {
            1 => IpProtocol::ICMP,
            6 => IpProtocol::TCP,
            17 => IpProtocol::UDP,
            58 => IpProtocol::ICMPv6,
            _ => IpProtocol::Other(value),
        }
    }
}

/// IPv4 packet header
#[derive(Debug, Clone)]
pub struct Ipv4Header {
    pub version: u8,
    pub ihl: u8,
    pub dscp: u8,
    pub ecn: u8,
    pub total_length: u16,
    pub identification: u16,
    pub flags: u8,
    pub fragment_offset: u16,
    pub ttl: u8,
    pub protocol: IpProtocol,
    pub checksum: u16,
    pub source: Ipv4Addr,
    pub destination: Ipv4Addr,
}

impl Ipv4Header {
    /// Minimum header size (20 bytes)
    pub const MIN_SIZE: usize = 20;

    /// Parse IPv4 header from bytes
    pub fn parse(data: &[u8]) -> Result<Self> {
        if data.len() < Self::MIN_SIZE {
            return Err(VpnError::InvalidPacket("IPv4 header too short".to_string()));
        }

        let version = (data[0] >> 4) & 0x0F;
        if version != 4 {
            return Err(VpnError::InvalidPacket(format!("Not IPv4: version={}", version)));
        }

        let ihl = data[0] & 0x0F;
        let header_len = (ihl as usize) * 4;

        if data.len() < header_len {
            return Err(VpnError::InvalidPacket("IPv4 header length mismatch".to_string()));
        }

        Ok(Self {
            version,
            ihl,
            dscp: (data[1] >> 2) & 0x3F,
            ecn: data[1] & 0x03,
            total_length: u16::from_be_bytes([data[2], data[3]]),
            identification: u16::from_be_bytes([data[4], data[5]]),
            flags: (data[6] >> 5) & 0x07,
            fragment_offset: u16::from_be_bytes([data[6] & 0x1F, data[7]]),
            ttl: data[8],
            protocol: IpProtocol::from(data[9]),
            checksum: u16::from_be_bytes([data[10], data[11]]),
            source: Ipv4Addr::new(data[12], data[13], data[14], data[15]),
            destination: Ipv4Addr::new(data[16], data[17], data[18], data[19]),
        })
    }

    /// Get header length in bytes
    pub fn header_len(&self) -> usize {
        (self.ihl as usize) * 4
    }

    /// Serialize header to bytes
    pub fn to_bytes(&self) -> Vec<u8> {
        let mut bytes = Vec::with_capacity(Self::MIN_SIZE);

        bytes.push((self.version << 4) | self.ihl);
        bytes.push((self.dscp << 2) | self.ecn);
        bytes.extend_from_slice(&self.total_length.to_be_bytes());
        bytes.extend_from_slice(&self.identification.to_be_bytes());
        bytes.push((self.flags << 5) | ((self.fragment_offset >> 8) as u8 & 0x1F));
        bytes.push((self.fragment_offset & 0xFF) as u8);
        bytes.push(self.ttl);
        bytes.push(self.protocol as u8);
        bytes.extend_from_slice(&self.checksum.to_be_bytes());
        bytes.extend_from_slice(&self.source.octets());
        bytes.extend_from_slice(&self.destination.octets());

        bytes
    }
}

/// VPN packet wrapper
#[derive(Debug, Clone)]
pub enum VpnPacket {
    /// IPv4 packet
    IPv4(Ipv4Packet),
    /// Handshake packet
    Handshake(HandshakePacket),
    /// Data packet (encrypted)
    Data(DataPacket),
}

/// IPv4 packet with payload
#[derive(Debug, Clone)]
pub struct Ipv4Packet {
    pub header: Ipv4Header,
    pub payload: Vec<u8>,
}

impl Ipv4Packet {
    /// Parse an IPv4 packet from raw bytes
    pub fn parse(data: &[u8]) -> Result<Self> {
        let header = Ipv4Header::parse(data)?;
        let header_len = header.header_len();
        let payload = data[header_len..header.total_length as usize].to_vec();

        Ok(Self { header, payload })
    }

    /// Serialize packet to bytes
    pub fn to_bytes(&self) -> Vec<u8> {
        let mut bytes = self.header.to_bytes();
        bytes.extend_from_slice(&self.payload);
        bytes
    }
}

/// Handshake packet for key exchange
#[derive(Debug, Clone)]
pub struct HandshakePacket {
    /// Packet type (1 = initiation, 2 = response)
    pub packet_type: u8,
    /// Sender's public key
    pub public_key: [u8; 32],
    /// Encrypted payload
    pub encrypted_payload: Vec<u8>,
}

impl HandshakePacket {
    /// Packet size constant
    pub const INITIATION_SIZE: usize = 148; // type(1) + reserved(3) + key(32) + encrypted(96) + mac(16)

    /// Parse handshake packet
    pub fn parse(data: &[u8]) -> Result<Self> {
        if data.len() < 4 {
            return Err(VpnError::InvalidPacket("Handshake packet too short".to_string()));
        }

        let packet_type = data[0];
        if packet_type != 1 && packet_type != 2 {
            return Err(VpnError::InvalidPacket(format!("Invalid handshake type: {}", packet_type)));
        }

        let mut public_key = [0u8; 32];
        public_key.copy_from_slice(&data[4..36]);

        let encrypted_payload = data[36..].to_vec();

        Ok(Self {
            packet_type,
            public_key,
            encrypted_payload,
        })
    }

    /// Serialize handshake packet
    pub fn to_bytes(&self) -> Vec<u8> {
        let mut bytes = Vec::with_capacity(4 + 32 + self.encrypted_payload.len());
        bytes.push(self.packet_type);
        bytes.extend_from_slice(&[0, 0, 0]); // reserved
        bytes.extend_from_slice(&self.public_key);
        bytes.extend_from_slice(&self.encrypted_payload);
        bytes
    }
}

/// Data packet (encrypted payload)
#[derive(Debug, Clone)]
pub struct DataPacket {
    /// Receiver index
    pub receiver_index: u32,
    /// Counter for nonce generation
    pub counter: u64,
    /// Encrypted payload
    pub encrypted_payload: Vec<u8>,
}

impl DataPacket {
    /// Parse data packet
    pub fn parse(data: &[u8]) -> Result<Self> {
        if data.len() < 16 {
            return Err(VpnError::InvalidPacket("Data packet too short".to_string()));
        }

        let receiver_index = u32::from_be_bytes([data[0], data[1], data[2], data[3]]);
        let counter = u64::from_be_bytes([
            data[4], data[5], data[6], data[7],
            data[8], data[9], data[10], data[11],
        ]);
        let encrypted_payload = data[12..].to_vec();

        Ok(Self {
            receiver_index,
            counter,
            encrypted_payload,
        })
    }

    /// Serialize data packet
    pub fn to_bytes(&self) -> Vec<u8> {
        let mut bytes = Vec::with_capacity(12 + self.encrypted_payload.len());
        bytes.extend_from_slice(&self.receiver_index.to_be_bytes());
        bytes.extend_from_slice(&self.counter.to_be_bytes());
        bytes.extend_from_slice(&self.encrypted_payload);
        bytes
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_ipv4_header_parse() {
        // Create a minimal IPv4 header
        let data = vec![
            0x45, 0x00, // version=4, IHL=5, DSCP=0, ECN=0
            0x00, 0x1C, // total length = 28
            0x00, 0x01, // identification = 1
            0x40, 0x00, // flags=0x4 (DF), fragment offset=0
            0x40, 0x11, // TTL=64, protocol=UDP (17)
            0x00, 0x00, // checksum = 0 (placeholder)
            0x0A, 0x00, 0x00, 0x01, // source = 10.0.0.1
            0x0A, 0x00, 0x00, 0x02, // destination = 10.0.0.2
        ];

        let header = Ipv4Header::parse(&data).unwrap();
        assert_eq!(header.version, 4);
        assert_eq!(header.ihl, 5);
        assert_eq!(header.total_length, 28);
        assert_eq!(header.protocol, IpProtocol::UDP);
        assert_eq!(header.source, Ipv4Addr::new(10, 0, 0, 1));
        assert_eq!(header.destination, Ipv4Addr::new(10, 0, 0, 2));
    }

    #[test]
    fn test_ipv4_header_serialize() {
        let header = Ipv4Header {
            version: 4,
            ihl: 5,
            dscp: 0,
            ecn: 0,
            total_length: 28,
            identification: 1,
            flags: 0x4,
            fragment_offset: 0,
            ttl: 64,
            protocol: IpProtocol::UDP,
            checksum: 0,
            source: Ipv4Addr::new(10, 0, 0, 1),
            destination: Ipv4Addr::new(10, 0, 0, 2),
        };

        let bytes = header.to_bytes();
        assert_eq!(bytes.len(), 20);

        // Parse it back
        let parsed = Ipv4Header::parse(&bytes).unwrap();
        assert_eq!(parsed.source, header.source);
        assert_eq!(parsed.destination, header.destination);
    }

    #[test]
    fn test_handshake_packet() {
        let packet = HandshakePacket {
            packet_type: 1,
            public_key: [0xAA; 32],
            encrypted_payload: vec![0xBB; 64],
        };

        let bytes = packet.to_bytes();
        let parsed = HandshakePacket::parse(&bytes).unwrap();

        assert_eq!(parsed.packet_type, 1);
        assert_eq!(parsed.public_key, [0xAA; 32]);
        assert_eq!(parsed.encrypted_payload, vec![0xBB; 64]);
    }

    #[test]
    fn test_data_packet() {
        let packet = DataPacket {
            receiver_index: 12345,
            counter: 67890,
            encrypted_payload: vec![0xCC; 100],
        };

        let bytes = packet.to_bytes();
        let parsed = DataPacket::parse(&bytes).unwrap();

        assert_eq!(parsed.receiver_index, 12345);
        assert_eq!(parsed.counter, 67890);
        assert_eq!(parsed.encrypted_payload, vec![0xCC; 100]);
    }
}

File: projects/vpn-software/vpn/packet.py
"""
VPN packet parsing and construction.

Handles:
- IPv4/IPv6 packet parsing
- VPN protocol packets (handshake, data, keepalive)
- Packet serialization/deserialization
"""

import struct
import socket
from enum import IntEnum
from dataclasses import dataclass, field
from typing import Optional, List

from vpn.error import InvalidPacketError, ProtocolError


# ---- Protocol Constants ----

class MessageType(IntEnum):
    """VPN protocol message types."""
    HANDSHAKE_INIT = 1
    HANDSHAKE_RESPONSE = 2
    TRANSPORT_DATA = 3
    KEEPALIVE = 4
    AUTH_REQUEST = 5
    AUTH_RESPONSE = 6
    DISCONNECT = 7


HEADER_MAGIC = 0x56504E01  # "VPN\x01"
PROTOCOL_VERSION = 1

# Packet sizes
HANDSHAKE_INIT_SIZE = 4 + 1 + 1 + 32 + 32 + 16  # magic + ver + type + pubkey + ephemeral + mac
HANDSHAME_RESPONSE_SIZE = 4 + 1 + 1 + 32 + 32 + 16
TRANSPORT_HEADER_SIZE = 4 + 1 + 1 + 4 + 8  # magic + ver + type + session_id + counter
AUTH_REQUEST_HEADER_SIZE = 4 + 1 + 1 + 4  # magic + ver + type + session_id


# ---- IPv4 Packet ----

@dataclass
class IPv4Header:
    """Parsed IPv4 header."""
    version: int = 4
    ihl: int = 5
    dscp: int = 0
    ecn: int = 0
    total_length: int = 0
    identification: int = 0
    flags: int = 0
    fragment_offset: int = 0
    ttl: int = 64
    protocol: int = 17  # UDP
    checksum: int = 0
    source: str = "0.0.0.0"
    destination: str = "0.0.0.0"

    MIN_SIZE = 20

    @staticmethod
    def parse(data: bytes) -> "IPv4Header":
        """Parse an IPv4 header from raw bytes."""
        if len(data) < IPv4Header.MIN_SIZE:
            raise InvalidPacketError("IPv4 header too short")

        version_ihl = data[0]
        version = (version_ihl >> 4) & 0x0F
        if version != 4:
            raise InvalidPacketError(f"Not IPv4: version={version}")

        ihl = version_ihl & 0x0F
        header_len = ihl * 4

        if len(data) < header_len:
            raise InvalidPacketError("IPv4 header length mismatch")

        dscp_ecn = data[1]
        total_length = struct.unpack("!H", data[2:4])[0]
        identification = struct.unpack("!H", data[4:6])[0]
        flags_frag = struct.unpack("!H", data[6:8])[0]
        flags = (flags_frag >> 13) & 0x07
        fragment_offset = flags_frag & 0x1FFF
        ttl = data[8]
        protocol = data[9]
        checksum = struct.unpack("!H", data[10:12])[0]
        source = socket.inet_ntoa(data[12:16])
        destination = socket.inet_ntoa(data[16:20])

        return IPv4Header(
            version=version,
            ihl=ihl,
            dscp=(dscp_ecn >> 2) & 0x3F,
            ecn=dscp_ecn & 0x03,
            total_length=total_length,
            identification=identification,
            flags=flags,
            fragment_offset=fragment_offset,
            ttl=ttl,
            protocol=protocol,
            checksum=checksum,
            source=source,
            destination=destination,
        )

    def to_bytes(self) -> bytes:
        """Serialize header to bytes."""
        dscp_ecn = ((self.dscp & 0x3F) << 2) | (self.ecn & 0x03)
        flags_frag = ((self.flags & 0x07) << 13) | (self.fragment_offset & 0x1FFF)
        src = socket.inet_aton(self.source)
        dst = socket.inet_aton(self.destination)

        header = struct.pack(
            "!BBHHHBBHII",
            (self.version << 4) | (self.ihl & 0x0F),
            dscp_ecn,
            self.total_length,
            self.identification,
            flags_frag,
            self.ttl,
            self.protocol,
            self.checksum,
            int.from_bytes(src, "big"),
            int.from_bytes(dst, "big"),
        )
        return header

    @property
    def header_length(self) -> int:
        return self.ihl * 4


@dataclass
class IPv4Packet:
    """A parsed IPv4 packet."""
    header: IPv4Header
    payload: bytes = b""

    @staticmethod
    def parse(data: bytes) -> "IPv4Packet":
        """Parse a complete IPv4 packet."""
        header = IPv4Header.parse(data)
        payload = data[header.header_length:header.total_length]
        return IPv4Packet(header=header, payload=payload)

    def to_bytes(self) -> bytes:
        """Serialize the packet."""
        return self.header.to_bytes() + self.payload

    @property
    def destination(self) -> str:
        return self.header.destination

    @property
    def source(self) -> str:
        return self.header.source


# ---- VPN Protocol Packets ----

@dataclass
class HandshakeInitPacket:
    """Handshake initiation packet."""
    sender_index: int = 0
    public_key: bytes = b""
    ephemeral_key: bytes = b""
    mac: bytes = b""

    def to_bytes(self) -> bytes:
        """Serialize the handshake initiation."""
        header = struct.pack(
            "!IBBH",
            HEADER_MAGIC,
            PROTOCOL_VERSION,
            MessageType.HANDSHAKE_INIT,
            self.sender_index,
        )
        return header + self.public_key + self.ephemeral_key + self.mac

    @staticmethod
    def parse(data: bytes) -> "HandshakeInitPacket":
        """Parse a handshake initiation packet."""
        if len(data) < 8 + 32 + 32:
            raise ProtocolError("Handshake init packet too short")

        magic = struct.unpack("!I", data[0:4])[0]
        if magic != HEADER_MAGIC:
            raise ProtocolError(f"Invalid magic: {magic:#x}")

        version = data[4]
        if version != PROTOCOL_VERSION:
            raise ProtocolError(f"Unsupported version: {version}")

        msg_type = data[5]
        if msg_type != MessageType.HANDSHAKE_INIT:
            raise ProtocolError(f"Expected handshake init, got {msg_type}")

        sender_index = struct.unpack("!H", data[6:8])[0]
        public_key = data[8:40]
        ephemeral_key = data[40:72]
        mac = data[72:88] if len(data) >= 88 else b""

        return HandshakeInitPacket(
            sender_index=sender_index,
            public_key=public_key,
            ephemeral_key=ephemeral_key,
            mac=mac,
        )


@dataclass
class HandshakeResponsePacket:
    """Handshake response packet."""
    sender_index: int = 0
    receiver_index: int = 0
    public_key: bytes = b""
    ephemeral_key: bytes = b""
    mac: bytes = b""

    def to_bytes(self) -> bytes:
        header = struct.pack(
            "!IBBHH",
            HEADER_MAGIC,
            PROTOCOL_VERSION,
            MessageType.HANDSHAKE_RESPONSE,
            self.sender_index,
            self.receiver_index,
        )
        return header + self.public_key + self.ephemeral_key + self.mac

    @staticmethod
    def parse(data: bytes) -> "HandshakeResponsePacket":
        if len(data) < 10 + 32 + 32:
            raise ProtocolError("Handshake response too short")

        magic = struct.unpack("!I", data[0:4])[0]
        if magic != HEADER_MAGIC:
            raise ProtocolError(f"Invalid magic: {magic:#x}")

        msg_type = data[5]
        if msg_type != MessageType.HANDSHAKE_RESPONSE:
            raise ProtocolError(f"Expected handshake response, got {msg_type}")

        sender_index = struct.unpack("!H", data[6:8])[0]
        receiver_index = struct.unpack("!H", data[8:10])[0]
        public_key = data[10:42]
        ephemeral_key = data[42:74]
        mac = data[74:90] if len(data) >= 90 else b""

        return HandshakeResponsePacket(
            sender_index=sender_index,
            receiver_index=receiver_index,
            public_key=public_key,
            ephemeral_key=ephemeral_key,
            mac=mac,
        )


@dataclass
class TransportDataPacket:
    """Encrypted transport data packet."""
    session_id: int = 0
    counter: int = 0
    encrypted_payload: bytes = b""

    HEADER_SIZE = TRANSPORT_HEADER_SIZE

    def to_bytes(self) -> bytes:
        header = struct.pack(
            "!IBBIQ",
            HEADER_MAGIC,
            PROTOCOL_VERSION,
            MessageType.TRANSPORT_DATA,
            self.session_id,
            self.counter,
        )
        return header + self.encrypted_payload

    @staticmethod
    def parse(data: bytes) -> "TransportDataPacket":
        if len(data) < TRANSPORT_HEADER_SIZE:
            raise ProtocolError("Transport data packet too short")

        magic = struct.unpack("!I", data[0:4])[0]
        if magic != HEADER_MAGIC:
            raise ProtocolError(f"Invalid magic: {magic:#x}")

        msg_type = data[5]
        if msg_type != MessageType.TRANSPORT_DATA:
            raise ProtocolError(f"Expected transport data, got {msg_type}")

        session_id = struct.unpack("!I", data[6:10])[0]
        counter = struct.unpack("!Q", data[10:18])[0]
        encrypted_payload = data[18:]

        return TransportDataPacket(
            session_id=session_id,
            counter=counter,
            encrypted_payload=encrypted_payload,
        )


@dataclass
class AuthRequestPacket:
    """Authentication request packet (username/password or certificate)."""
    session_id: int = 0
    auth_type: int = 0  # 1=password, 2=certificate
    username: str = ""
    password_hash: bytes = b""
    certificate: bytes = b""

    def to_bytes(self) -> bytes:
        username_bytes = self.username.encode("utf-8")
        header = struct.pack(
            "!IBBIB",
            HEADER_MAGIC,
            PROTOCOL_VERSION,
            MessageType.AUTH_REQUEST,
            self.session_id,
            self.auth_type,
        )
        if self.auth_type == 1:
            # Password auth
            return (
                header
                + struct.pack("!H", len(username_bytes))
                + username_bytes
                + self.password_hash
            )
        else:
            # Certificate auth
            return (
                header
                + struct.pack("!H", len(username_bytes))
                + username_bytes
                + struct.pack("!I", len(self.certificate))
                + self.certificate
            )

    @staticmethod
    def parse(data: bytes) -> "AuthRequestPacket":
        if len(data) < 11:
            raise ProtocolError("Auth request too short")

        magic = struct.unpack("!I", data[0:4])[0]
        if magic != HEADER_MAGIC:
            raise ProtocolError(f"Invalid magic: {magic:#x}")

        msg_type = data[5]
        if msg_type != MessageType.AUTH_REQUEST:
            raise ProtocolError(f"Expected auth request, got {msg_type}")

        session_id = struct.unpack("!I", data[6:10])[0]
        auth_type = data[10]
        offset = 11

        username_len = struct.unpack("!H", data[offset:offset + 2])[0]
        offset += 2
        username = data[offset:offset + username_len].decode("utf-8")
        offset += username_len

        if auth_type == 1:
            password_hash = data[offset:offset + 32]
            return AuthRequestPacket(
                session_id=session_id,
                auth_type=auth_type,
                username=username,
                password_hash=password_hash,
            )
        else:
            cert_len = struct.unpack("!I", data[offset:offset + 4])[0]
            offset += 4
            certificate = data[offset:offset + cert_len]
            return AuthRequestPacket(
                session_id=session_id,
                auth_type=auth_type,
                username=username,
                certificate=certificate,
            )


@dataclass
class AuthResponsePacket:
    """Authentication response packet."""
    session_id: int = 0
    success: bool = False
    message: str = ""

    def to_bytes(self) -> bytes:
        msg_bytes = self.message.encode("utf-8")
        header = struct.pack(
            "!IBBIB",
            HEADER_MAGIC,
            PROTOCOL_VERSION,
            MessageType.AUTH_RESPONSE,
            self.session_id,
            1 if self.success else 0,
        )
        return header + struct.pack("!H", len(msg_bytes)) + msg_bytes

    @staticmethod
    def parse(data: bytes) -> "AuthResponsePacket":
        if len(data) < 11:
            raise ProtocolError("Auth response too short")

        magic = struct.unpack("!I", data[0:4])[0]
        if magic != HEADER_MAGIC:
            raise ProtocolError(f"Invalid magic: {magic:#x}")

        session_id = struct.unpack("!I", data[6:10])[0]
        success = data[10] == 1
        msg_len = struct.unpack("!H", data[11:13])[0]
        message = data[13:13 + msg_len].decode("utf-8")

        return AuthResponsePacket(
            session_id=session_id,
            success=success,
            message=message,
        )


@dataclass
class KeepalivePacket:
    """Keepalive packet."""
    session_id: int = 0

    def to_bytes(self) -> bytes:
        return struct.pack(
            "!IBBI",
            HEADER_MAGIC,
            PROTOCOL_VERSION,
            MessageType.KEEPALIVE,
            self.session_id,
        )

    @staticmethod
    def parse(data: bytes) -> "KeepalivePacket":
        if len(data) < 10:
            raise ProtocolError("Keepalive packet too short")
        session_id = struct.unpack("!I", data[6:10])[0]
        return KeepalivePacket(session_id=session_id)


def parse_message_type(data: bytes) -> MessageType:
    """Parse just the message type from raw bytes."""
    if len(data) < 6:
        raise ProtocolError("Packet too short")
    magic = struct.unpack("!I", data[0:4])[0]
    if magic != HEADER_MAGIC:
        raise ProtocolError(f"Invalid magic: {magic:#x}")
    return MessageType(data[5])
